tests/jobs/unit/aws_iot_tests_jobs_api.c

```diff
@@ -682,6 +695,68 @@ TEST( Jobs_Unit_API, RemovePersistentSubscriptions )
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief Tests the behavior of the Jobs callback functions.
+ */
+TEST( Jobs_Unit_API, SetCallback )
+{
+    AwsIotJobsError_t status = AWS_IOT_JOBS_STATUS_PENDING;
+    AwsIotJobsCallbackInfo_t callbackInfo = AWS_IOT_JOBS_CALLBACK_INFO_INITIALIZER;
+    _jobsSubscription_t * pSubscription = NULL;
+
+    /* Callback info must be provided. */
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, TEST_THING_NAME, TEST_THING_NAME_LENGTH, 0, NULL );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_BAD_PARAMETER, status );
+
+    /* Thing Name must be valid. */
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, NULL, 0, 0, &callbackInfo );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_BAD_PARAMETER, status );
+
+    /* Request to remove an unspecified callback. */
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, TEST_THING_NAME, TEST_THING_NAME_LENGTH, 0, &callbackInfo );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_BAD_PARAMETER, status );
+
+    /* Request to remove a callback that doesn't exist. */
+    callbackInfo.function = NULL;
+    callbackInfo.oldFunction = JOBS_CALLBACK_FUNCTION;
+
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, TEST_THING_NAME, TEST_THING_NAME_LENGTH, 0, &callbackInfo );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_SUCCESS, status );
+
+    /* Set new callback. */
+    callbackInfo.function = JOBS_CALLBACK_FUNCTION;
+    callbackInfo.oldFunction = NULL;
+
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, TEST_THING_NAME, TEST_THING_NAME_LENGTH, 0, &callbackInfo );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_SUCCESS, status );
+
+    /* Check that new callback was set. */
+    pSubscription = _AwsIotJobs_FindSubscription( TEST_THING_NAME, TEST_THING_NAME_LENGTH, false );
+    TEST_ASSERT_NOT_NULL( pSubscription );
+    TEST_ASSERT_EQUAL_PTR( pSubscription->callbacks[ NOTIFY_NEXT_CALLBACK ].function, JOBS_CALLBACK_FUNCTION );
+
+    /* Replace existing function. */
+    callbackInfo.function = JOBS_CALLBACK_FUNCTION_2;
+    callbackInfo.oldFunction = JOBS_CALLBACK_FUNCTION;
+
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, TEST_THING_NAME, TEST_THING_NAME_LENGTH, 0, &callbackInfo );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_SUCCESS, status );
+
+    /* Check that callback was replaced. */
+    pSubscription = _AwsIotJobs_FindSubscription( TEST_THING_NAME, TEST_THING_NAME_LENGTH, false );
+    TEST_ASSERT_NOT_NULL( pSubscription );
+    TEST_ASSERT_EQUAL_PTR( pSubscription->callbacks[ NOTIFY_NEXT_CALLBACK ].function, JOBS_CALLBACK_FUNCTION_2 );
+
+    /* Remove callback function. */
+    callbackInfo.function = NULL;
+    callbackInfo.oldFunction = JOBS_CALLBACK_FUNCTION_2;
+
+    status = AwsIotJobs_SetNotifyNextCallback( _pMqttConnection, TEST_THING_NAME, TEST_THING_NAME_LENGTH, 0, &callbackInfo );
+    TEST_ASSERT_EQUAL( AWS_IOT_JOBS_SUCCESS, status );
```

Comment:
Check that callback did get removed by checking the that _findSubscription returns NULL.